src/client/Client.cc

```diff
@@ -7571,29 +7584,38 @@ int Client::_do_setattr(Inode *in, struct ceph_statx *stx, int mask,
 force_request:
   MetaRequest *req = new MetaRequest(CEPH_MDS_OP_SETATTR);
 
-  filepath path;
-
-  in->make_nosnap_relative_path(path);
-  req->set_filepath(path);
-  req->set_inode(in);
-
   if (mask & CEPH_SETATTR_KILL_SGUID) {
     req->inode_drop |= CEPH_CAP_AUTH_SHARED;
   }
   if (mask & CEPH_SETATTR_MODE) {
-    req->head.args.setattr.mode = stx->stx_mode;
-    req->inode_drop |= CEPH_CAP_AUTH_SHARED;
-    ldout(cct,10) << "changing mode to " << stx->stx_mode << dendl;
+    if (!in->caps_issued_mask(CEPH_CAP_AUTH_SHARED) ||
+        stx->stx_mode != in->mode) {
```

Comment:
It's pretty confusing having the logic for each CEPH_SETATTR_* value split into two different stanzas -- one that occurs before "force request:" and one that occurs after. It sort of makes sense in the existing code -- we check to see if we can buffer the attribute change and if we can't we then build a request and send it off.

This patch changes that though and it's harder to follow the logic around whether the change will be buffered.

Why not do the testing/handling for As caps and the mode change in the CEPH_SETATTR_MODE section above this point (around line #7524)?

Ditto for the other CEPH_SETATTR_* changes.

Reply:
> It's pretty confusing having the logic for each CEPH_SETATTR_* value split into two different stanzas -- one that occurs before "force request:" and one that occurs after. It sort of makes sense in the existing code -- we check to see if we can buffer the attribute change and if we can't we then build a request and send it off.
> 

Yeah, this sounds good.

> This patch changes that though and it's harder to follow the logic around whether the change will be buffered.
> 
> Why not do the testing/handling for As caps and the mode change in the CEPH_SETATTR_MODE section above this point (around line #7524)?
> 
> Ditto for the other CEPH_SETATTR_* changes.

Sure, will improve it.